narwhals/_pyspark/expr.py

```diff
@@ -0,0 +1,213 @@
+    ) -> Self:
+        def func(df: PySparkLazyFrame) -> list[Column]:
+            from pyspark.sql import functions as F  # noqa: N812
+
+            _ = df
+            return [F.col(col_name) for col_name in column_names]
+
+        return cls(
+            func,
+            depth=0,
+            function_name="col",
+            root_names=list(column_names),
+            output_names=list(column_names),
+            returns_scalar=False,
+            backend_version=backend_version,
+            dtypes=dtypes,
+        )
+
+    def _from_call(
+        self,
+        call: Callable[..., Column],
+        expr_name: str,
+        *args: PySparkExpr,
+        returns_scalar: bool,
+        **kwargs: PySparkExpr,
+    ) -> Self:
+        def func(df: PySparkLazyFrame) -> list[Column]:
+            results = []
+            inputs = self._call(df)
+            _args = [maybe_evaluate(df, arg) for arg in args]
+            _kwargs = {key: maybe_evaluate(df, value) for key, value in kwargs.items()}
+            for _input in inputs:
+                input_col_name = get_column_name(df, _input)
+                column_result = call(_input, *_args, **_kwargs)
+                if not returns_scalar:
+                    column_result = column_result.alias(input_col_name)
+                results.append(column_result)
+            return results
+
+        # Try tracking root and output names by combining them from all
+        # expressions appearing in args and kwargs. If any anonymous
+        # expression appears (e.g. nw.all()), then give up on tracking root names
+        # and just set it to None.
+        root_names = copy(self._root_names)
+        output_names = self._output_names
+        for arg in list(args) + list(kwargs.values()):
+            if root_names is not None and isinstance(arg, self.__class__):
+                if arg._root_names is not None:
+                    root_names.extend(arg._root_names)
+                else:  # pragma: no cover
+                    root_names = None
+                    output_names = None
+                    break
+            elif root_names is None:
+                output_names = None
+                break
+
+        if not (
+            (output_names is None and root_names is None)
+            or (output_names is not None and root_names is not None)
+        ):  # pragma: no cover
+            msg = "Safety assertion failed, please report a bug to https://github.com/narwhals-dev/narwhals/issues"
+            raise AssertionError(msg)
+
+        return self.__class__(
+            func,
+            depth=self._depth + 1,
+            function_name=f"{self._function_name}->{expr_name}",
+            root_names=root_names,
+            output_names=output_names,
+            returns_scalar=self._returns_scalar or returns_scalar,
+            backend_version=self._backend_version,
+            dtypes=self._dtypes,
+        )
+
+    def __add__(self, other: PySparkExpr) -> Self:
```

Comment:
Would other require to be parsed into column (e.g. if it is an expr)? Same for all other dunder methods